### PR TITLE
[TF-7737] Add fields for Registry Module Testing

### PR DESCRIPTION
**Pull request by hashimoon**
## Description

These changes are to support testing functionality for the Private Module Registry. 
- Receive the `TestConfig` object on the `RegistryModule`
- Enable tests for a new `RegistryModule` by passing a `TestConfig` object with `TestsEnabled` set to `true`
- Enable tests for an existing `RegistryModule` by passing a `TestConfig` object with `TestsEnabled` set to `true`
- Enable switching a `RegistryModule` between `git_tag` and `branch` based publishing

## Output from tests
Including output from tests may require access to a TFE instance. Ignore this section if you have no environment to test against.

```
$ TFE_ADDRESS="https://example" TFE_TOKEN="example" go test ./... -v -run TestFunctionsAffectedByChange


=== RUN   TestRegistryModuleUpdatePrivateRegistry
=== RUN   TestRegistryModuleUpdatePrivateRegistry/enable_no-code
2023/09/29 12:39:31 [WARN] Support for using the NoCode field is deprecated as of release 1.22.0 and may be removed in a future version. The preferred way to update a no-code module is with the registryNoCodeModules.Update method.
=== RUN   TestRegistryModuleUpdatePrivateRegistry/disable_no-code
2023/09/29 12:39:31 [WARN] Support for using the NoCode field is deprecated as of release 1.22.0 and may be removed in a future version. The preferred way to update a no-code module is with the registryNoCodeModules.Update method.
=== RUN   TestRegistryModuleUpdatePrivateRegistry/toggle_between_git_tag-based_and_branch-based_publishing
--- PASS: TestRegistryModuleUpdatePrivateRegistry (5.97s)
    --- PASS: TestRegistryModuleUpdatePrivateRegistry/enable_no-code (0.30s)
    --- PASS: TestRegistryModuleUpdatePrivateRegistry/disable_no-code (0.25s)
    --- PASS: TestRegistryModuleUpdatePrivateRegistry/toggle_between_git_tag-based_and_branch-based_publishing (0.84s)
PASS
ok  	github.com/hashicorp/go-tfe	6.419s

=== RUN   TestRegistryModulesCreateBranchBasedWithVCSConnectionWithTesting
=== RUN   TestRegistryModulesCreateBranchBasedWithVCSConnectionWithTesting/with_valid_options
=== RUN   TestRegistryModulesCreateBranchBasedWithVCSConnectionWithTesting/with_valid_options/tests_are_enabled
=== RUN   TestRegistryModulesCreateBranchBasedWithVCSConnectionWithTesting/with_invalid_options
=== RUN   TestRegistryModulesCreateBranchBasedWithVCSConnectionWithTesting/with_invalid_options/when_the_the_module_is_not_branch_based_and_test_are_enabled
--- PASS: TestRegistryModulesCreateBranchBasedWithVCSConnectionWithTesting (5.88s)
    --- PASS: TestRegistryModulesCreateBranchBasedWithVCSConnectionWithTesting/with_valid_options (2.51s)
        --- PASS: TestRegistryModulesCreateBranchBasedWithVCSConnectionWithTesting/with_valid_options/tests_are_enabled (0.00s)
    --- PASS: TestRegistryModulesCreateBranchBasedWithVCSConnectionWithTesting/with_invalid_options (0.00s)
        --- PASS: TestRegistryModulesCreateBranchBasedWithVCSConnectionWithTesting/with_invalid_options/when_the_the_module_is_not_branch_based_and_test_are_enabled (0.00s)
PASS
ok  	github.com/hashicorp/go-tfe	6.233s
...
```

### CODE REVIEW

File: registry_module.go
Comment (brandonc) [inline registry_module.go:718]: Unless I am mistaken, I think it's possible for this field to be nil. I think that would cause a nil pointer dereference panic.

File: test_config.go
Comment (brandonc) [inline test_config.go:3]: I think I would prefer if this struct were defined alongside `RegistryModuleTestConfigOptions`. WDYT?

Comment (brandonc) [reply test_config.go:3]: And perhaps this could have a similar name since TestConfig is rather generic

File: registry_module.go
Comment (nfagerlund) [reply registry_module.go:718]: Good catch. Yeah, this needs a nil check please!

File: registry_module.go
Comment (nfagerlund) [inline registry_module.go:267]: These fields are unexpected, because there's only one Update method, which acts on both VCS-backed and non-VCS-backed modules. 

Also, I can't find any API docs for the PATCH endpoint with registry modules, so I'm not sure how this corresponds to the backing API yet. 🤔 If you've got the details, that'd be helpful to know.

What's the expected behavior when you provide one of these for a non-VCS module? What's the expected behavior when you provide _both_ of these? I see you're not doing any validity checking in the client in this PR, so presumably you get whatever the backend gives ya...

At minimum these fields need some godoc comments to clarify when and how to use em.

File: test_config.go
Comment (nfagerlund) [reply test_config.go:3]: Agreed. Actually, I'd like a bit of context here: 

- Do we predict other objects in TFC will have test configs of some kind? (providers, workspaces...?)
- Do we predict test configs will have additional options beyond "on / off"? (I'm guessing yes, since you went to the trouble of creating a struct, but let's lay it out explicitly while we're reviewing here!) 

Prior art: the VCSRepo struct lives over in workspace.go, even though a couple of different types nest it.

File: registry_module.go
Comment (nfagerlund) [inline registry_module.go:134]: Within the client, the traditional pattern for "magic string" API fields (where logically speaking it's an enum, it's just that none of our languages handle enums very usefully) is to define a new string type and then define constants for all of the legal values of the field. See the RegistryModuleStatus type right here in this file for an example. 

I'd like to see that for this field, unless there's a good reason to do something different.

**Review by nfagerlund — CHANGES_REQUESTED** [review]

File: test_config.go
Comment (hashimoon) [reply test_config.go:3]: > Agreed. Actually, I'd like a bit of context here:
> 
> * Do we predict other objects in TFC will have test configs of some kind? (providers, workspaces...?)
Yes - the backend was with that in mind.

> * Do we predict test configs will have additional options beyond "on / off"? (I'm guessing yes, since you went to the trouble of creating a struct, but let's lay it out explicitly while we're reviewing here!)
Yes - although we are not supporting any additional options at the moment. The separate struct is also reflective of there being a separate Rails model/database table on the backend.
> 
> Prior art: the VCSRepo struct lives over in workspace.go, even though a couple of different types nest it.

I'm happy to move the struct wherever you feel is best.

File: registry_module.go
Comment (hashimoon) [reply registry_module.go:267]: I didn't add any client side validation because the update method only takes the `RegistryModuleID`, which does not feature the `PublishingMechanism`, which is what would be needed to enforce client side validation on these fields. 

I can add additional documentation on how these fields operate.

File: registry_module.go
Comment (uturunku1) [reply registry_module.go:734]: addressed!

**Review by uturunku1 — APPROVED** [review]

**Comment by uturunku1** [comment]
I'm dismissing Nick's request for changes as is blocking the PR from being merged. They are not available today. But I have taken look this PR to made sure that Joseph addressed the changes requested by Nick.

File: registry_module.go
Comment (sebasslash) [inline registry_module.go:273]:
```suggestion
	VCSRepo *RegistryModuleVCSRepoUpdateOptions `jsonapi:"attr,vcs-repo,omitempty"`
```

**Review by sebasslash — CHANGES_REQUESTED** [review]
Some test failures I got locally:
```
2023/10/06 13:23:11 [WARN] Support for using the NoCode field is deprecated as of release 1.22.0 and may be removed in a future version. The preferred way to create a no-code module is with the registryNoCodeModules.Create method.
    registry_module_integration_test.go:157:
                Error Trace:    /Users/sebastianrivera/hashicorp/go-tfe/registry_module_integration_test.go:157
                Error:          Received unexpected error:
                                resource not found
                Test:           TestRegistryModulesCreate/with_valid_options/with_no-code_attribute
```

```
 2023/10/06 13:23:11 [WARN] Support for using the NoCode field is deprecated as of release 1.22.0 and may be removed in a future version. The preferred way to create a no-code module is with the registryNoCodeModules.Create method.
    registry_module_integration_test.go:157:
                Error Trace:    /Users/sebastianrivera/hashicorp/go-tfe/registry_module_integration_test.go:157
                Error:          Received unexpected error:
                                resource not found
                Test:           TestRegistryModulesCreate/with_valid_options/with_no-code_attribute
```

```
 2023/10/06 13:23:13 [WARN] Support for using the NoCode field is deprecated as of release 1.22.0 and may be removed in a future version. The preferred way to update a no-code module is with the registryNoCodeModules.Update method.
    registry_module_integration_test.go:283:
                Error Trace:    /Users/sebastianrivera/hashicorp/go-tfe/registry_module_integration_test.go:283
                Error:          Received unexpected error:
                                Bad jsonapi struct tag format
                Test:           TestRegistryModuleUpdate/enable_no-code
=== RUN   TestRegistryModuleUpdate/disable_no-code
2023/10/06 13:23:13 [WARN] Support for using the NoCode field is deprecated as of release 1.22.0 and may be removed in a future version. The preferred way to update a no-code module is with the registryNoCodeModules.Update method.
    registry_module_integration_test.go:298:
                Error Trace:    /Users/sebastianrivera/hashicorp/go-tfe/registry_module_integration_test.go:298
                Error:          Received unexpected error:
                                Bad jsonapi struct tag format
                Test:           TestRegistryModuleUpdate/disable_no-code
```

**Pull request merged**
